tests/src/test/scala/whisk/core/containerpool/test/ContainerPoolTests.scala

```diff
@@ -0,0 +1,384 @@
+class ContainerPoolTests extends TestKit(ActorSystem("ContainerPool"))
+    with ImplicitSender
+    with FlatSpecLike
+    with Matchers
+    with BeforeAndAfterAll
+    with MockFactory {
+
+    override def afterAll = TestKit.shutdownActorSystem(system)
+
+    val timeout = 5.seconds
+
+    // Common entities to pass to the tests. We don't really care what's inside
+    // those for the behavior testing here, as none of the contents will really
+    // reach a container anyway. We merely assert that passing and extraction of
+    // the values is done properly.
+    val exec = CodeExecAsString(RuntimeManifest("actionKind"), "testCode", None)
+    val memoryLimit = 256.MB
+
+    val invocationNamespace = EntityName("invocationSpace")
+    val action = ExecutableWhiskAction(EntityPath("actionSpace"), EntityName("actionName"), exec)
+
+    val message = ActivationMessage(
+        TransactionId.testing,
+        action.fullyQualifiedName(true),
+        action.rev,
+        Identity(Subject(), invocationNamespace, AuthKey(), Set()),
+        ActivationId(),
+        invocationNamespace.toPath,
+        None)
+
+    val runMessage = Run(action, message)
+
+    /** Helper to create PreWarmedData */
+    def preWarmedData(kind: String, memoryLimit: ByteSize = memoryLimit) = PreWarmedData(stub[Container], kind, memoryLimit)
+
+    /** Helper to create WarmedData */
+    def warmedData(action: ExecutableWhiskAction = action, namespace: String = "invocationSpace", lastUsed: Instant = Instant.now) =
+        WarmedData(stub[Container], EntityName(namespace), action, lastUsed)
+
+    /** Creates a sequence of containers and a factory returning this sequence. */
+    def testContainers(n: Int) = {
+        val containers = (0 to n).map(_ => TestProbe())
+        val queue = mutable.Queue(containers: _*)
+        val factory = (fac: ActorRefFactory) => queue.dequeue().ref
+        (containers, factory)
+    }
+
+    behavior of "ContainerPool"
+
+    it should "indicate free resources to the feed only if a warm container responds" in within(timeout) {
+        val (containers, factory) = testContainers(1)
+        val feed = TestProbe()
+
+        val pool = system.actorOf(ContainerPool.props(factory, 0, feed.ref))
+        containers(0).send(pool, NeedWork(warmedData()))
+        feed.expectMsg(ContainerReleased)
+    }
+
+    /*
+     * CONTAINER SCHEDULING
+     *
+     * These tests only test the simplest approaches. Look below for full coverage tests
+     * of the respective scheduling methods.
+     */
+    it should "reuse a warm container" in within(timeout) {
+        val (containers, factory) = testContainers(2)
+        val feed = TestProbe()
+        val pool = system.actorOf(ContainerPool.props(factory, 2, feed.ref))
+
+        pool ! runMessage
+        containers(0).expectMsg(runMessage)
+        containers(0).send(pool, NeedWork(warmedData()))
+
+        pool ! runMessage
+        containers(0).expectMsg(runMessage)
```

Comment:
I can, but that'll have the test run until the timeout in every case. Maybe I can mitigate and only wait for 200ms for noMessages.